Objects/fileobject.c

```diff
@@ -2356,7 +2349,30 @@ file_iternext(PyFileObject *f)
     if (!f->readable)
         return err_mode("reading");
 
-    l = readahead_get_line_skip(f, 0, READAHEAD_BUFSIZE);
+    {
+        /*
+          Multiple threads can enter this method while the GIL is released
+          during file read and wreak havoc on the file object's readahead
+          buffer. To avoid dealing with cross-thread coordination issues, we
+          cache the file buffer state locally and only set it back on the file
+          object when we're done.
+        */
+        readaheadbuffer rab = {f->f_buf, f->f_bufptr, f->f_bufend};
```

Comment:
drop_readaheadbuffer() frees `rab->buf` which is `f->f_buf`. If two threads call drop_readaheadbuffer() simultaneously `f->f_buf` will be freed twice.

Reply:
Other threads can't access `f->f_buf` because it's set to `NULL` for the duration of the method.